File: topics/leetcode/easy/121_best_time_to_buy_and_sell_stock.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        cheapest_cost = prices[0]

        for price in prices:
            if price < cheapest_cost:
                cheapest_cost = price
            if price - cheapest_cost > max_profit:
                max_profit = price - cheapest_cost

        return max_profit
    

# for this problem we have an O(n) solution.
# the main thing to remember is that we can only sell stock in the future
# so all we need to do is keep track of the lowest price we have seen
# and the greatest difference we will find.

File: topics/leetcode/easy/121_best_time_to_buy_and_sell_stock.cpp
#include <vector>
using namespace std;

class Solution {
    public:
        int maxProfit(vector<int>& prices) {
            int max_profit = 0;
            int least = prices[0];
            for (int i = 0; i < prices.size(); i++) {
                if (prices[i] < least)
                    least = prices[i];
                if ((prices[i] - least) > max_profit)
                    max_profit = prices[i] - least;
            }
            return max_profit;
        }
};